### PR TITLE
Use constant generics for specialized `memcmp`

### PR DESCRIPTION
**Pull request by zakcutner**
I noticed that we can now use constant generics for our specialized `memcmp` implementations which removes a lot of unsafe code (particularly in a place where we fixed a bug before). I tried it in Godbolt and it seems the compiler can automatically perform the specialization that we're currently doing manually.

I'm actually seeing a small performance improvement by making this change as I assume the compiler does something cleverer than we were. I also removed the test cases for `memcmp` since we are only now calling a single library function but I can add that back in if you think.

### CODE REVIEW

File: src/memcmp.rs [hunk 1]
```diff
@@ -1,173 +1,11 @@
...
-
-#[inline]
-pub unsafe fn memcmp12(left: *const u8, right: *const u8, n: usize) -> bool {
-    debug_assert_eq!(n, 12);
-    memcmp8(left, right, 8) && memcmp4(left.add(8), right.add(8), 4)
-}
-
-#[inline]
-pub unsafe fn memcmp(left: *const u8, right: *const u8, n: usize) -> bool {
+pub unsafe fn unspecialized(left: *const u8, right: *const u8, n: usize) -> bool {
```
Comment (marmeladema) [inline src/memcmp.rs:4]: Maybe `generic` would be a better name?

**Comment by marmeladema** [comment]
Codewise the change looks great, but I will need to run the perf benchmarks locally (and maybe internally) before approving this :)

File: src/memcmp.rs (hunk 1 above)
Comment (zakcutner) [reply src/memcmp.rs:4]: That's a good suggestion, I'll update that 👍

**Comment by zakcutner** [comment]
Thanks! That's definitely a good idea, I'm not sure how much I trust the benchmarks on my Mac in any case 😄

**Comment by marmeladema** [comment]
Here are the results of the perf benchmarks:
* `master` branch:
```
long_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1574173 instructions 1574173 instructions 1574173 instructions]
long_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1574173 instructions 1574173 instructions 1574173 instructions]
short_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1008880107 instructions 1008880107 instructions 1008880108 instructions]
short_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1008880092 instructions 1008880092 instructions 1008880093 instructions]
```
* `memcmp` branch:
```
long_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1574317 instructions 1574317 instructions 1574317 instructions]
long_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1574317 instructions 1574317 instructions 1574317 instructions]
short_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1008871636 instructions 1008871636 instructions 1008871637 instructions]
short_haystack/DynamicAvx2Searcher::search_in #2
                        time:   [1008871637 instructions 1008871638 instructions 1008871639 instructions]
```

Summary:
* `long_haystack/DynamicAvx2Searcher::search_in`: +0.00914766039056758%
* `short_haystack/DynamicAvx2Searcher::search_in`: -0.0008379588483364842%

**Review by marmeladema — APPROVED** [review]

**Comment by zakcutner** [comment]
@marmeladema Cool, thanks for checking this! I saw you also approved the PR, do you want to run some internal benchmarks first or should we merge?

**Comment by marmeladema** [comment]
Let's wait until I run the internal benchmarks on Monday.

**Pull request merged**